blockchain/consensus/narwhal_cns/src/exec.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright 2019-2022 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+
+use crate::consensus::NarwhalConsensusError;
+use async_std::channel::Sender;
+use async_trait::async_trait;
+use forest_blocks::{Error as BlockchainError, Ticket, Tipset};
+use forest_chain::ChainStore;
+use forest_chain::Error as ChainStoreError;
+use forest_crypto::VRFProof;
+use forest_encoding::tuple::*;
+use forest_ipld_blockstore::BlockStore;
+use fvm_ipld_encoding::Cbor;
+use narwhal_consensus::ConsensusOutput;
+use narwhal_executor::{
+    BatchExecutionState, ExecutionState, ExecutionStateError, SerializedTransaction,
+};
+use narwhal_types::SequenceNumber;
+
+pub struct NarwhalOutput {
+    pub consensus_output: ConsensusOutput,
+    pub transaction_batches: Vec<Vec<SerializedTransaction>>,
+}
+
+pub struct NarwhalExecutionState<BS> {
+    chain_store: Arc<ChainStore<BS>>,
+    output_tx: Sender<NarwhalOutput>,
+    locked: AtomicBool,
+}
+
+impl<BS> NarwhalExecutionState<BS> {
+    pub fn new(chain_store: Arc<ChainStore<BS>>, output_tx: Sender<NarwhalOutput>) -> Self {
+        Self {
+            chain_store,
+            output_tx,
+            locked: AtomicBool::new(false),
+        }
+    }
+}
+
+impl ExecutionStateError for NarwhalConsensusError {
+    fn node_error(&self) -> bool {
+        // For now treat all errors as fatal.
+        true
+    }
+}
+
+impl<BS> ExecutionState for NarwhalExecutionState<BS> {
+    type Error = NarwhalConsensusError;
+
+    fn ask_consensus_write_lock(&self) -> bool {
+        self.locked
+            .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
+            .is_ok()
+    }
+
+    fn release_consensus_write_lock(&self) {
+        self.locked.store(false, Ordering::Relaxed)
+    }
+}
+
+#[async_trait]
+impl<BS> BatchExecutionState for NarwhalExecutionState<BS>
+where
+    BS: BlockStore + Sync + Send + 'static,
+{
+    /// Load the tip of the chain and get the last certificate index from the `Ticket` in the header.
+    async fn load_next_certificate_index(&self) -> Result<SequenceNumber, Self::Error> {
+        match self.chain_store.heaviest_tipset().await {
+            None => Ok(0),
+            Some(tipset) => {
+                let index = ConsensusTransactionIndex::try_from(tipset.as_ref())?;
+                Ok(index.next_consensus_index())
+            }
+        }
+    }
+
+    async fn handle_consensus(
+        &self,
+        consensus_output: &ConsensusOutput,
+        transaction_batches: Vec<Vec<SerializedTransaction>>,
+    ) -> Result<(), Self::Error> {
+        // Append the transactions and the certificate to a pending queue that we can take values from
```

Comment:
I think this is the right approach (at least for now, until we understand better how we want to handle the syncing process). The TipsetProcessor notifies the syncer about new heads available so that it can proactively fetch new blocks if it sees that its current local view is out of date for several blocks (I think this is also used to handle potential forks in EC).

Reply:
Yes the tipset processor expects to sync ranges, and fetch gaps. Which would be good for non-validators to catch up with validators. But in the local case that we are dealing with here, there will never be gaps, so all this not required and I should be able to just generate a block, save it, append it to the chain as the heaviest tipset and execute it, something like [this](https://github.com/aakoshh/forest/blob/main/blockchain/chain_sync/src/tipset_syncer.rs#L988-L989). It would probably work as well, because it is the `ChainStore` that notifies subscribers that need to know (at least I think so). 

This is where our abstractions are so leaky: once we have a reference to the `StateManager` we have access to _everything_. 

If we were to think about how to do a "clean" start, we should think in more granular interfaces: what are the operations (some) consensus wants a ledger to provide, like stake weights, and allow different ledger/consensus combinations to be composed together without passing in concrete implementations like the `StateManager` that assumes certain actors will be present and gives you access to a chain store that assumes tipsets and also gives you access to IPLD and raw KV storage.